### PR TITLE
T2D-250 Split Reference Sequence to allow multiple genes with single analysis

### PR DESCRIPTION
**Pull request by iyangar**
Entity ReferenceSequence is split to Assemby, Gene and Transcriptome allowing only M:M with Analysis and Gene.
M:1 continues with Analysis and Assembly/Transcriptome.
Changes:
- Dependencies are updated
- Test cases updated
- Minor changes to swagger
- ReferenceSequence specific tests like Type is removed

### CODE REVIEW

File: src/main/java/uk/ac/ebi/ampt2d/metadata/SpringDataRestConfig.java [hunk 1]
```diff
@@ -32,24 +32,14 @@
...
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
 import uk.ac.ebi.ampt2d.metadata.aop.StudyDeprecationAspect;
-import uk.ac.ebi.ampt2d.metadata.persistence.entities.Analysis;
-import uk.ac.ebi.ampt2d.metadata.persistence.entities.ReferenceSequence;
-import uk.ac.ebi.ampt2d.metadata.persistence.entities.File;
-import uk.ac.ebi.ampt2d.metadata.persistence.entities.Sample;
-import uk.ac.ebi.ampt2d.metadata.persistence.entities.Study;
-import uk.ac.ebi.ampt2d.metadata.persistence.entities.Taxonomy;
-import uk.ac.ebi.ampt2d.metadata.persistence.entities.WebResource;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.*;
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/SpringDataRestConfig.java]: In our Java coding style guide, we specifically disallow wildcard import: https://www.ebi.ac.uk/seqdb/confluence/display/VAR/Java+coding+style#Javacodingstyle-Importstatements

You can disable the wildcard import automatically imposed by the IDE like the answers in this SO issue: https://stackoverflow.com/questions/3348816/intellij-never-use-wildcard-imports

Same for other places.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/SpringDataRestConfig.java [hunk 2]
```diff
@@ -80,7 +70,9 @@ public void configureRepositoryRestConfiguration(RepositoryRestConfiguration con
                 config.getCorsRegistry().addMapping("/**").allowedMethods("*").allowedOrigins("*");
                 config.exposeIdsFor(
                         Analysis.class,
-                        ReferenceSequence.class,
+                        Assembly.class,
+                        Transcriptome.class,
+                        Gene.class,
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/SpringDataRestConfig.java]: Maybe it is just me, by `Gene` I am thinking more about something else than the reference gene sequence used in an analysis. But I don't have a better name for this class.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/GeneRestController.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Autowired
+    private GeneRepository geneRepository;
+
+    @Autowired
+    private GenericResourceAssembler<Gene, GeneResource> resourceAssembler;
+
+    @ApiOperation(value="Get a filtered list of genes based on filtering criteria")
+    @ApiImplicitParams({
+            @ApiImplicitParam(name = "name", value = "name", dataType = "string", paramType = "query", example = "GRCh38"),
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/rest/controllers/GeneRestController.java]: `GRCh38` is an assembly name. Please use a real world gene name as an example.

Same for other places.

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 4]
```diff
@@ -118,34 +120,60 @@ public void cleanDatabases() throws Exception {
...
+                .andExpect(status().isCreated()).andReturn();
+
+        return mvcResult.getResponse().getHeader("Location");
     }
 
-    private void postReferenceSequence(String name, String patch, List accessions, ReferenceSequence.Type type) throws Exception {
-        String location = postTestReferenceSequence(name, patch, accessions, type);
+    @Test
+    public void postTranscriptome() throws Exception {
+        String name = "GRCh37";
```
Comment (Zhicheng-Liu) [inline src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: `GRCh37` is an assembly name. Can you use a name that does not look like a gene or an assembly, or even better a real world transcriptome name.

Same for other places.

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 5]
```diff
@@ -227,32 +255,43 @@ private String postTestStudy(String accession, int version, String name, String
...
+        String geneUrl = postTestGene("GRCh39", "p4",
+                Arrays.asList("GCA_000001405.5", "GCF_000001405.16"));
+        geneUrlList.add(geneUrl);
+        geneUrl = postTestGene("GRCh40", "p5",
+                Arrays.asList("GCA_000001405.5", "GCF_000001405.16"));
+        geneUrlList.add(geneUrl);
         String studyUrl = postTestStudy("EGAS0001", 1, "test_human_study");
 
-        String location = postTestAnalysis("EGAA0001", referenceSequenceUrl, studyUrl, Analysis.Technology.GWAS,
+        String location = postTestAnalysis("EGAA0001", assemblyUrl, transcriptomeUrl, geneUrlList, studyUrl, Analysis.Technology.GWAS,
```
Comment (Zhicheng-Liu) [inline src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: Can you also test the cases in which an analysis is based on only one type of reference sequence? In fact, I think for now we should disallow an analysis based on three different types of reference sequence. @cyenyxe

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java [hunk 6]
```diff
@@ -90,26 +90,36 @@
...
-    private ReferenceSequence referenceSequence;
+    private Assembly assembly;
 
-    @ApiModelProperty(position = 6, required = true)
+    @ApiModelProperty(position = 6, dataType = "java.lang.String", notes = "Url to a Transcriptome")
+    @JsonProperty
+    @ManyToOne(optional = false)
+    private Transcriptome transcriptome;
+
+    @ApiModelProperty(position = 7, dataType = "java.lang.String", example = "[url1, url2]", notes = "Url to Gene List")
```
Comment (selvaebi) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java]: notes should be  "List of Gene Urls"

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java (hunk 5 above)
Comment (iyangar) [reply src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: I will verify if T2D-218 changes can help here.

File: metadata-ws/src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 7]
```diff
@@ -247,28 +247,293 @@ public void postAnalysis() throws Exception {
...
+
+        List<String> referenceSequenceList = new ArrayList<>();
+        referenceSequenceList.add(referenceSequenceUrl1);
+        referenceSequenceList.add(referenceSequenceUrl2);
+        String location = postTestAnalysis("EGAA0001", referenceSequenceList, studyUrl, Analysis.Technology.GWAS,
+                Analysis.Type.CASE_CONTROL, "Illumina");
+
+        mockMvc.perform(get(location))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.accessionVersionId.accession").value("EGAA0001"));
```
Comment (selvaebi) [inline metadata-ws/src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: perform check of reference sequence here before the update as well.

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/exceptionhandling/AnalysisWithoutReferenceSequenceException.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ */
+package uk.ac.ebi.ampt2d.metadata.exceptionhandling;
+
+public class AnalysisWithoutReferenceSequenceException extends IllegalArgumentException {
+
+    private static final long serialVersionUID = 5740339279937775231L;
+
+    public AnalysisWithoutReferenceSequenceException() {
+        super("An analysis must have at least one reference sequence");
```
Comment (cyenyxe) [inline metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/exceptionhandling/AnalysisWithoutReferenceSequenceException.java]:
```suggestion
        super("An analysis must be mapped against one or more reference sequences");
```

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/exceptionhandling/InvalidReferenceSequenceException.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ */
+package uk.ac.ebi.ampt2d.metadata.exceptionhandling;
+
+public class InvalidReferenceSequenceException extends IllegalArgumentException {
+
+    private static final long serialVersionUID = 7071006298434627630L;
+
+    public InvalidReferenceSequenceException() {
+        super("At least one of the reference sequence URL is invalid");
```
Comment (cyenyxe) [inline metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/exceptionhandling/InvalidReferenceSequenceException.java]:
```suggestion
        super("At least one of the reference sequence URLs is invalid");
```

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java [hunk 10]
```diff
@@ -102,10 +102,11 @@
     @ManyToOne(optional = false)
     private Study study;
 
-    @ApiModelProperty(position = 6, dataType = "java.lang.String", notes = "Url to an Reference Sequence")
+    @ApiModelProperty(position = 6, dataType = "java.lang.String", example = "[url1, url2]",
+            notes = "A single reference sequence URL of assembly/transcriptome type alone or reference sequences URL list of gene type")
```
Comment (cyenyxe) [inline metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java]:
```suggestion
            notes = "URL(s) to the reference sequence(s): either a single URL to an assembly/transcriptome, or a comma-separated list of gene sequences")
```

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    public void validateReferenceSequences(Analysis analysis, Object referenceSequence) {
+        validateReferenceSequenceLink(analysis);
+    }
+
+    private void validateReferenceSequenceLink(Analysis analysis) {
+        if (analysis.getReferenceSequences() == null || analysis.getReferenceSequences().size() == 0) {
+            throw new AnalysisWithoutReferenceSequenceException();
+        } else if (analysis.getReferenceSequences().contains(null)) {
+            throw new InvalidReferenceSequenceException();
+        } else if (analysis.getReferenceSequences().size() >= 2) {
```
Comment (cyenyxe) [inline metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java]: I personally find more clear to use

```suggestion
        } else if (analysis.getReferenceSequences().size() > 1) {
```

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    @HandleBeforeLinkDelete
+    @HandleBeforeLinkSave
+    public void validateReferenceSequences(Analysis analysis, Object referenceSequence) {
+        validateReferenceSequenceLink(analysis);
+    }
+
+    private void validateReferenceSequenceLink(Analysis analysis) {
+        if (analysis.getReferenceSequences() == null || analysis.getReferenceSequences().size() == 0) {
+            throw new AnalysisWithoutReferenceSequenceException();
+        } else if (analysis.getReferenceSequences().contains(null)) {
```
Comment (cyenyxe) [inline metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java]: When could this happen?

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+            throw new InvalidReferenceSequenceException();
+        } else if (analysis.getReferenceSequences().size() >= 2) {
+            boolean invalidRefSeq = analysis.getReferenceSequences().stream().anyMatch(r -> r.getType().equals(ReferenceSequence.Type.ASSEMBLY) ||
+                    r.getType().equals(ReferenceSequence.Type.TRANSCRIPTOME));
```
Comment (cyenyxe) [inline metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java]: I think the condition would express better the intention if written the other way around:

```java
boolean invalidRefSeq = analysis.getReferenceSequences().stream().anyMatch(r -> !r.getType().equals(ReferenceSequence.Type.GENE));
```

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+            if (invalidRefSeq) {
+                throw new InvalidReferenceSequenceException("Invalid type of reference sequences. " +
+                        "If the reference sequence URL list is more than one, all of them should be of gene type");
```
Comment (cyenyxe) [inline metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java]:
```suggestion
                        "When multiple reference sequence URLs are provided, all of them should point to gene sequences");
```

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java (hunk 12 above)
Comment (iyangar) [reply metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java]: https://github.com/EBIvariation/metadata-ws/blob/4d78893f6a97abfcc9936516578f53c040138b78/metadata-ws/src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java#L424

https://github.com/EBIvariation/metadata-ws/blob/4d78893f6a97abfcc9936516578f53c040138b78/metadata-ws/src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java#L461

File: metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java (hunk 14 above)
Comment (iyangar) [reply metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/events/AnalysisEventHandler.java]: All changes are completed

**Pull request merged**